doc: Add a page about dqlite troubleshooting (#13346)
canonical/lxd: doc/howto/dqlite_troubleshoot.md

> # How to troubleshoot (some) dqlite errors

Reviewer: ```suggestion
# How to troubleshoot (some) Dqlite errors
```
This is a global change.

> dqlite is the distributed database that LXD uses to store information
> that must be synchronized across a cluster. This page describes how
> to respond if LXD fails to start or crashes because of a dqlite-related error.

Reviewer: ```suggestion
dqlite is the distributed database that LXD uses to store information
that must be synchronized across a cluster. 
See {ref}`database` for more information.

This how-to guide describes strategies for how to respond to Dqlite-related errors.
```

## Recognizing dqlite-related errors

> Look for error messages that mention keywords like `dqlite`, `raft`, or
> `segment`, especially if LXD is failing to start up.

Reviewer: ```suggestion
If LXD fails to start up or crashes, you should suspect a Dqlite-related error if the error message mentions keywords like `dqlite`, `raft`, or `segment`.
```

Reviewer: Moving the symptoms (failing to start or crashing) into this section from the intro.

A known risk factor for some of the errors covered below is a previous LXD
crash caused by running out of disk space.

## The dqlite data directory

When investigating dqlite-related errors, it's essential to look at the
contents of the dqlite data directory for the affected node. This is the
directory where the local instance of dqlite stores all its data. If you
installed the LXD snap, you can find this directory at
`/var/snap/lxd/common/lxd/database/global`.

The data directory contains several types of file. The most important types are:

> - Closed segments: these have filenames like
>   `0000000000056436-0000000000056501`. The two numbers are the *start index*
>    and *end index*. Both indices are inclusive.
> - Open segments: these have filenames like `open-1`
> - Snapshot files: these have names like `snapshot-1-59392-27900`. The first
>   number is the *snapshot index*.
> - Snapshot metadata files: these have names like `snapshot-1-59392-27900.meta`
>   and are paired with snapshot files.

Reviewer: ```suggestion
- Closed segments: These have filenames like
  `0000000000056436-0000000000056501`. The two numbers are the *start index*
   and *end index*. Both indices are inclusive.
- Open segments: These have filenames like `open-1`.
- Snapshot files: These have names like `snapshot-1-59392-27900`. The first
  number is the *snapshot index*.
- Snapshot metadata files: These have names like `snapshot-1-59392-27900.meta`
  and are paired with snapshot files.
```

> ## Spotting weirdness

Reviewer: ```suggestion
## Spotting unexpected segments
```

Or similar. 
I like the weirdness, but not in documentation. :smile:

Author: Is "anomalies" better?

Reviewer: Yes :)

When looking at the contents of the data directory, watch for the following
symptoms:

> 1. Closed segments whose index ranges overlap (remember that these ranges are inclusive)
> 2. A closed segment with end index X where the next closed segment has start index greater than X + 1
> 3. A snapshot file with snapshot index X where the next closed segment has start index greater than X + 1
> 4. A snapshot file whose size is less than the size of a previous (lower-numbered) snapshot

Reviewer: ```suggestion
- Closed segments whose index ranges overlap (remember that these ranges are inclusive).
- A closed segment with end index X where the next closed segment has start index greater than X + 1.
- A snapshot file with snapshot index X where the next closed segment has start index greater than X + 1.
- A snapshot file whose size is less than the size of a previous (lower-numbered) snapshot.
```

I'm actually not sure why the linter complains about the ordered list prefixes - but is there a reason to use an ordered list here? This seems like a list of things to look out for, no matter in which order.

Author: The reason is just to be able to refer unambigously (by number) to the symptoms in the next section. I could use an unordered list and just say put "(symptom A)" (etc.) after each item?

Author: It's a weird markdownlint quirk, see https://github.com/canonical/lxd/pull/13346/commits/df765b97e69d6dc7a4cbb4c43f64c03a6843f283

When scanning for these symptoms, start with the most recent snapshots and
closed segments (those with the highest indices) since the problem is more
likely to be there.

## Specific error messages

> - `closed segment [...] is past last snapshot [...]`: this indicates that you have symptom (3) above, possibly combined with symptom (1)
> - `load closed segment [...]: entries count in preamble is zero`: this indicates that the mentioned segment is corrupt

Reviewer: ```suggestion
- `closed segment [...] is past last snapshot [...]`: 

  This indicates that you have symptom (3) above, possibly combined with symptom (1).
- `load closed segment [...]: entries count in preamble is zero`:

  This indicates that the mentioned segment is corrupt.
```

Reviewer: OK, I see you use the numbers here. Can you be more explicit though (something like "This indicates that a snapshot overlaps with ... something ;)")?

## Interventions

> **Important:** Before taking any of the actions below, back up the entire
> dqlite data directory, so you don't lose data in case something goes wrong.

Reviewer: ````suggestion
```{important}
Before taking any of the actions below, back up the entire
Dqlite data directory, so you don't lose data in case something goes wrong.
```
````

Here are some actions you can take in response to specific dqlite errors. They

> are aimed at restoring the invariants described above without destroying any

Reviewer: invariants?

Author: oops, legacy of a previous draft

data. They are not guaranteed to work in any specific case.

> - If you have overlapping closed segments (symptom 1), try deleting some of
>   them to remove the overlap, without creating gaps in the sequence of indices
>   or removing any index that was previously represented
> - If the snapshot file with the highest index is unexpectedly small (symptom
>   4), and there are still closed segments covering all the indices up to and
>   including this snapshot's index, delete the snapshot and its corresponding
>   metadata file

Reviewer: ```suggestion
- If you have overlapping closed segments (symptom 1), try deleting some of
  them to remove the overlap, without creating gaps in the sequence of indices
  or removing any index that was previously represented.
- If the snapshot file with the highest index is unexpectedly small (symptom
  4), and there are still closed segments covering all the indices up to and
  including this snapshot's index, delete the snapshot and its corresponding
  metadata file.
```

- If the last (highest-numbered) closed segment is corrupt, try deleting it.
  (Deleting closed segments before the last one will create a gap and generally
  prevent dqlite from starting.)

## Get help

If the tips above don't help with your situation, you can always post on the

> LXD support forum. Please mention dqlite in your post, tag @colemiller and

Reviewer: ```suggestion
LXD support forum. Please mention dqlite in your post, tag `@colemiller` and
```

Reviewer: Not sure I like mentioning specific user names in the docs ... do we have a group we could mention instead?

Reviewer 2: agreed, lets just reference the support forum

> include the error message(s) you're seeing, LXD logs, and the output of

Reviewer: ```suggestion
include the error message(s) you're seeing, LXD logs, and the output of the following command (if you're using the LXD snap):
```

```

> $ sudo ls -lah /var/snap/lxd/common/lxd/database/global

Reviewer: ```suggestion
sudo ls -lah /var/snap/lxd/common/lxd/database/global
```

```

> (or, if you're not using the LXD snap, wherever your dqlite data directory
> lives). Also mention any troubleshooting steps you've already taken and what

Reviewer: ```suggestion
Also mention any troubleshooting steps you've already taken and what
```
